// roarr/dist/factories/createLogger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createLogger = void 0;
const config_1 = require("../config");
const constants_1 = require("../constants");
const hasOwnProperty_1 = require("../utilities/hasOwnProperty");
const isBrowser_1 = require("../utilities/isBrowser");
const isTruthy_1 = require("../utilities/isTruthy");
const createMockLogger_1 = require("./createMockLogger");
const fast_printf_1 = require("fast-printf");
const safe_stable_stringify_1 = __importDefault(require("safe-stable-stringify"));
let loggedWarningAsyncLocalContext = false;
const getGlobalRoarrContext = () => {
    return globalThis.ROARR;
};
const defaultAsyncLocalContext = {
    messageContext: {},
    transforms: [],
};
const getAsyncLocalContext = () => {
    const asyncLocalStorage = getGlobalRoarrContext().asyncLocalStorage;
    if (!asyncLocalStorage) {
        throw new Error('AsyncLocalContext is unavailable.');
    }
    const asyncLocalContext = asyncLocalStorage.getStore();
    if (asyncLocalContext) {
        return asyncLocalContext;
    }
    return defaultAsyncLocalContext;
};
const isAsyncLocalContextAvailable = () => {
    return Boolean(getGlobalRoarrContext().asyncLocalStorage);
};
const getSequence = () => {
    if (isAsyncLocalContextAvailable()) {
        const asyncLocalContext = getAsyncLocalContext();
        if ((0, hasOwnProperty_1.hasOwnProperty)(asyncLocalContext, 'sequenceRoot') &&
            (0, hasOwnProperty_1.hasOwnProperty)(asyncLocalContext, 'sequence') &&
            typeof asyncLocalContext.sequence === 'number') {
            return (String(asyncLocalContext.sequenceRoot) +
                '.' +
                String(asyncLocalContext.sequence++));
        }
        return String(getGlobalRoarrContext().sequence++);
    }
    return String(getGlobalRoarrContext().sequence++);
};
const MAX_ONCE_ENTRIES = 1000;
const buildOnceKey = (logLevel, a, b) => {
    if (typeof a === 'string') {
        return `${logLevel}:${a}`;
    }
    try {
        return `${logLevel}:${JSON.stringify(a)}:${b}`;
    }
    catch (_a) {
        return `${logLevel}:${(0, safe_stable_stringify_1.default)(a)}:${b}`;
    }
};
// Shared prototype for all logger instances.
// Methods use `this` (the log function itself) to dispatch,
// avoiding per-instance closure allocations.
const loggerPrototype = Object.create(Function.prototype);
for (const logLevelName of Object.keys(constants_1.logLevels)) {
    const logLevel = constants_1.logLevels[logLevelName];
    loggerPrototype[logLevelName] = function (a, b, c, d, e, f, g, h, index, index_) {
        if (typeof a === 'string') {
            this({ logLevel }, a, b, c, d, e, f, g, h, index);
        }
        else {
            this({ ...a, logLevel }, b, c, d, e, f, g, h, index, index_);
        }
    };
    loggerPrototype[logLevelName + 'Once'] = function (a, b, c, d, e, f, g, h, index, index_) {
        const onceLog = getGlobalRoarrContext().onceLog;
        const key = buildOnceKey(logLevel, a, b);
        if (onceLog.has(key)) {
            return;
        }
        onceLog.add(key);
        if (onceLog.size > MAX_ONCE_ENTRIES) {
            onceLog.clear();
        }
        if (typeof a === 'string') {
            this({ logLevel }, a, b, c, d, e, f, g, h, index);
        }
        else {
            this({ ...a, logLevel }, b, c, d, e, f, g, h, index, index_);
        }
    };
}
loggerPrototype.child = function (context) {
    const onMessage = this.onMessage;
    const parentMessageContext = this.parentMessageContext;
    const transforms = this.transforms;
    let asyncLocalContext;
    if (isAsyncLocalContextAvailable()) {
        asyncLocalContext = getAsyncLocalContext();
    }
    else {
        asyncLocalContext = defaultAsyncLocalContext;
    }
    if (typeof context === 'function') {
        // eslint-disable-next-line @typescript-eslint/no-use-before-define
        return (0, exports.createLogger)(onMessage, {
            ...asyncLocalContext.messageContext,
            ...parentMessageContext,
            ...context,
        }, [context, ...transforms]);
    }
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return (0, exports.createLogger)(onMessage, {
        ...asyncLocalContext.messageContext,
        ...parentMessageContext,
        ...context,
    }, transforms);
};
loggerPrototype.getContext = function () {
    const parentMessageContext = this.parentMessageContext;
    let asyncLocalContext;
    if (isAsyncLocalContextAvailable()) {
        asyncLocalContext = getAsyncLocalContext();
    }
    else {
        asyncLocalContext = defaultAsyncLocalContext;
    }
    return {
        ...asyncLocalContext.messageContext,
        ...parentMessageContext,
    };
};
loggerPrototype.adopt = async function (routine, context) {
    if (!isAsyncLocalContextAvailable()) {
        if (loggedWarningAsyncLocalContext === false) {
            loggedWarningAsyncLocalContext = true;
            this.onMessage({
                context: {
                    logLevel: constants_1.logLevels.warn,
                    package: 'roarr',
                },
                message: 'async_hooks are unavailable; Roarr.adopt will not function as expected',
                sequence: getSequence(),
                time: Date.now(),
                version: config_1.ROARR_LOG_FORMAT_VERSION,
            });
        }
        return routine();
    }
    const asyncLocalContext = getAsyncLocalContext();
    let sequenceRoot;
    if ((0, hasOwnProperty_1.hasOwnProperty)(asyncLocalContext, 'sequenceRoot') &&
        (0, hasOwnProperty_1.hasOwnProperty)(asyncLocalContext, 'sequence') &&
        typeof asyncLocalContext.sequence === 'number') {
        sequenceRoot =
            asyncLocalContext.sequenceRoot +
                '.' +
                String(asyncLocalContext.sequence++);
    }
    else {
        sequenceRoot = String(getGlobalRoarrContext().sequence++);
    }
    let nextContext = {
        ...asyncLocalContext.messageContext,
    };
    const nextTransforms = [...asyncLocalContext.transforms];
    if (typeof context === 'function') {
        nextTransforms.push(context);
    }
    else {
        nextContext = {
            ...nextContext,
            ...context,
        };
    }
    const asyncLocalStorage = getGlobalRoarrContext().asyncLocalStorage;
    if (!asyncLocalStorage) {
        throw new Error('Async local context unavailable.');
    }
    return asyncLocalStorage.run({
        messageContext: nextContext,
        sequence: 0,
        sequenceRoot,
        transforms: nextTransforms,
    }, () => {
        return routine();
    });
};
const createLogger = (onMessage, parentMessageContext = {}, transforms = []) => {
    var _a;
    if (!(0, isBrowser_1.isBrowser)() && typeof process !== 'undefined') {
        // eslint-disable-next-line node/no-process-env
        const enabled = (0, isTruthy_1.isTruthy)((_a = process.env.ROARR_LOG) !== null && _a !== void 0 ? _a : '');
        if (!enabled) {
            return (0, createMockLogger_1.createMockLogger)(onMessage, parentMessageContext);
        }
    }
    const log = (a, b, c, d, e, f, g, h, index, index_) => {
        var _a;
        const time = Date.now();
        const globalContext = globalThis.ROARR;
        const asyncLocalStorage = globalContext.asyncLocalStorage;
        const asyncLocalContext = (_a = asyncLocalStorage === null || asyncLocalStorage === void 0 ? void 0 : asyncLocalStorage.getStore()) !== null && _a !== void 0 ? _a : defaultAsyncLocalContext;
        let sequence;
        if ('sequenceRoot' in asyncLocalContext &&
            typeof asyncLocalContext.sequence === 'number') {
            sequence =
                asyncLocalContext.sequenceRoot +
                    '.' +
                    String(asyncLocalContext.sequence++);
        }
        else {
            sequence = String(globalContext.sequence++);
        }
        let context;
        let message;
        if (typeof a === 'string') {
            context = {
                ...asyncLocalContext.messageContext,
                ...parentMessageContext,
            };
        }
        else {
            context = {
                ...asyncLocalContext.messageContext,
                ...parentMessageContext,
                ...a,
            };
        }
        if (typeof a === 'string' && b === undefined) {
            message = a;
        }
        else if (typeof a === 'string') {
            if (!a.includes('%')) {
                throw new Error('When a string parameter is followed by other arguments, then it is assumed that you are attempting to format a message using printf syntax. You either forgot to add printf bindings or if you meant to add context to the log message, pass them in an object as the first parameter.');
            }
            message = (0, fast_printf_1.printf)(a, b, c, d, e, f, g, h, index, index_);
        }
        else {
            let fallbackMessage = b;
            if (typeof b !== 'string') {
                if (b === undefined) {
                    fallbackMessage = '';
                }
                else {
                    throw new TypeError('Message must be a string. Received ' + typeof b + '.');
                }
            }
            message = (0, fast_printf_1.printf)(fallbackMessage, c, d, e, f, g, h, index, index_);
        }
        let packet = {
            context,
            message,
            sequence,
            time,
            version: config_1.ROARR_LOG_FORMAT_VERSION,
        };
        if (asyncLocalContext.transforms.length > 0 || transforms.length > 0) {
            for (const transform of asyncLocalContext.transforms) {
                packet = transform(packet);
                if (typeof packet !== 'object' || packet === null) {
                    throw new Error('Message transform function must return a message object.');
                }
            }
            for (const transform of transforms) {
                packet = transform(packet);
                if (typeof packet !== 'object' || packet === null) {
                    throw new Error('Message transform function must return a message object.');
                }
            }
        }
        onMessage(packet);
    };
    log.onMessage = onMessage;
    log.parentMessageContext = parentMessageContext;
    log.transforms = transforms;
    Object.setPrototypeOf(log, loggerPrototype);
    return log;
};
exports.createLogger = createLogger;
